File: first non repeating character in a stream.cpp
// time: O(n) and space: O(26)

class Solution {
public:
	string FirstNonRepeating(string A) {
		// Code here
		unordered_map<char, int> seen;
		string result;
		queue<char> q;
		for (auto i : A) {
			q.push(i);
			if (seen.count(i)) seen[i] += 1;
			else seen[i] = 1;
			while (q.size() > 0 && seen[q.front()] > 1) {
				q.pop();
			}
			if (q.size() > 0) {
				result += q.front();
			}
			else {
				result += '#';
			}
		}
		return result;
	}

};

File: first non repeating character in a stream.py
from collections import deque
class Solution:
    def FirstNonRepeating(self, A):
        seen = {}
        q = deque()
        result = ""
        for char in A:
            q.append(char)
            if char in seen:
                seen[char] += 1
            else:
                seen[char] = 1
            while q and seen[q[0]] > 1:
                q.popleft()
            if q:
                result += q[0]
            else:
                result += '#'
        return result
